cpp/src/phonenumbers/phonenumberutil.cc

```diff
@@ -1527,12 +1525,76 @@ void PhoneNumberUtil::FormatNsnWithCarrier(const string& number,
       : metadata.intl_number_format();
   const NumberFormat* formatting_pattern =
       ChooseFormattingPatternForNumber(available_formats, number);
+  NumberFormat new_format;
+  bool consistent_optional_rule = TRUE;
```

Comment:
Why don't you just track `national_prefix_optional`?  Default to `TRUE`, then if any format says `FALSE`, set it to `FALSE`.